packages/flutter/lib/src/rendering/animated_size.dart

```diff
@@ -0,0 +1,100 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/animation.dart';
+
+import 'box.dart';
+import 'object.dart';
+import 'shifted_box.dart';
+
+/// A render object that can render child size change animations.
+class RenderAnimatedSize extends RenderAligningShiftedBox {
+  /// Creates a render animated size.
+  RenderAnimatedSize({
+    Duration duration,
+    Curve curve,
+    RenderBox child
+  }) : super(child: child, alignment: FractionalOffset.center) {
+    _controller = new AnimationController(
+      duration: duration
+    );
+    _animation = new CurvedAnimation(
+      parent: _controller,
+      curve: curve
+    );
+  }
+
+  AnimationController _controller;
+  CurvedAnimation _animation;
+  Size _oldSize;
+  Size _newSize;
+
+  /// The duration of the animation.
+  Duration get duration => _controller.duration;
+  set duration(Duration value) {
+    assert(value != null);
+    if (value == _controller.duration)
+      return;
+    _controller.duration = value;
+  }
+
+  /// The curve of the animation.
+  Curve get curve => _animation.curve;
+  set curve(Curve value) {
+    assert(value != null);
+    if (value == _animation.curve)
+      return;
+    _animation.curve = value;
+  }
+
+  @override
+  void detach() {
+    _controller?.stop();
+    super.detach();
+  }
+
+  void _trigger() {
+    _controller
+      ..removeListener(markNeedsLayout)
+      ..value = 0.0
+      ..forward()
+      ..addListener(markNeedsLayout);
+  }
+
+  Size _getSize() {
+    return Size.lerp(_oldSize, _newSize, _animation.value);
+  }
+
+  @override
+  void performLayout() {
+    if (child != null) {
+      child.layout(constraints, parentUsesSize: true);
+
+      if (_newSize == null) {
+        _oldSize = _newSize = constraints.constrain(child.size);
+      } else {
+        final Size childSize = constraints.constrain(child.size);
+
+        if (_newSize != childSize) {
+          _oldSize = _newSize;
+          _newSize = childSize;
+          _trigger();
+        }
+      }
+
+      size = _getSize();
+      alignChild();
+    } else {
+      size = _oldSize = _newSize = constraints.constrain(Size.zero);
+    }
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    if (child != null) {
+      final Rect rect = new Rect.fromLTWH(0.0, 0.0, size.width, size.height);
+      context.pushClipRect(needsCompositing, offset, rect, super.paint);
```

Comment:
You should only push a clip rect if the child actually visually overflows.  You should compute whether there is visual overflow in the `performLayout` function.  (Search in the rendering directory for "visualoverflow")